draft-ietf-quic-transport.md

```diff
@@ -2856,24 +2856,30 @@ needing acknowledgement are received.  The sender can use the receiver's
 Strategies and implications of the frequency of generating acknowledgments are
 discussed in more detail in {{QUIC-RECOVERY}}.
 
-To limit ACK Blocks to those that have not yet been received by the sender,
-the receiver SHOULD track which ACK frames have been acknowledged.  The
-receiver SHOULD exclude already acknowledged packets from future ACK frames
-whenever these packets would unnecessarily contribute to the ACK frame size.
+To limit the ranges of acknowledged packet numbers to those that have not yet
+been received by the sender, the receiver SHOULD track which ACK frames have
+been acknowledged by its peer.  The receiver SHOULD exclude already acknowledged
+packets from future ACK frames whenever these packets would unnecessarily
+contribute to the ACK frame size.
 
 Because ACK frames are not sent in response to ACK-only packets, a receiver that
 is only sending ACK frames will only receive acknowledgements for its packets if
 the sender includes them in packets with non-ACK frames.  A sender SHOULD bundle
 ACK frames with other frames when possible.
 
 To limit receiver state or the size of ACK frames, a receiver MAY limit the
-number of ACK Blocks it sends.  A receiver can do this even without receiving
+number of ACK Ranges it sends.  A receiver can do this even without receiving
 acknowledgment of its ACK frames, with the knowledge this could cause the sender
 to unnecessarily retransmit some data.  Standard QUIC {{QUIC-RECOVERY}}
 algorithms declare packets lost after sufficiently newer packets are
 acknowledged.  Therefore, the receiver SHOULD repeatedly acknowledge newly
 received packets in preference to packets received in the past.
 
+An endpoint SHOULD treat receipt of an acknowledgment for a packet it did not
+send as a connection error of type PROTOCOL_VIOLATION, if it is able to detect
```

Comment:
How so?  It should never happen...

Reply:
None of our errors (except for NO_ERROR) should ever happen. We have error codes for exactly those cases where things happen that are not supposed to happen.

Reply:
I don't have a strong opinion, but PROTOCOL_VIOLATION seems like a fine error for this.

Reply:
We had PROTOCOL_VIOLATIONs come up in interop testing a couple of times. The problem is, it tells you **exactly nothing** about what's going wrong.

gQUIC was (relatively) easy to debug, partly because it has a very long list of error codes. At some point (and I don't remember when and why), we decided to shrink that list to just around 10 error codes. Now we don't even have a single error code left for loss-recovery related errors - it's all just a PROTOCOL_VIOLATION.

Reply:
@marten-seemann Shouldn't we be using Reason Phrase for debugging? It provides the endpoint to decide how granular the error reports can be, without requiring all the stack handle the errors the same way.

Reply:
Actually, I think of the reason phrase as giving more details about the error, e.g. when exceeding the stream limit, I'd send the error code STREAM_LIMIT_ERROR, and then the reason phrase might say "Peer tried to open stream 124, but stream limit was 116.".

If we say that all information should go into the reason phrase, then why do we have FLOW_CONTROL_ERROR, STREAM_LIMIT_ERROR, FINAL_OFFSET_ERROR, FRAME_ENCODING_ERROR and TRANSPORT_PARAMETER_ERROR? They're certainly just protocol violations of one sort or another.

Reply:
Let's do a separate discussion about error codes. I don't think it's necessary here, and if this is an important new error code, let's introduce that in a new issue/PR.